File: impls/lua/step2_eval.lua
#!/usr/bin/env lua

local table = require('table')

local readline = require('readline')
local utils = require('utils')
local types = require('types')
local reader = require('reader')
local printer = require('printer')
local List, Vector, HashMap = types.List, types.Vector, types.HashMap

-- read
function READ(str)
    return reader.read_str(str)
end

-- eval

function EVAL(ast, env)

    -- print("EVAL: " .. printer._pr_str(ast, true))

    if types._symbol_Q(ast) then
        if env[ast.val] == nil then
            types.throw("'"..ast.val.."' not found")
        end
        return env[ast.val]
    elseif types._vector_Q(ast) then
        return Vector:new(utils.map(function(x) return EVAL(x,env) end,ast))
    elseif types._hash_map_Q(ast) then
        local new_hm = {}
        for k,v in pairs(ast) do
            new_hm[k] = EVAL(v, env)
        end
        return HashMap:new(new_hm)
    elseif not types._list_Q(ast) or #ast == 0 then
        return ast
    end

    local f = EVAL(ast[1], env)
    local args = types.slice(ast, 2)
    args = utils.map(function(x) return EVAL(x,env) end, args)
    return f(table.unpack(args))
end

-- print
function PRINT(exp)
    return printer._pr_str(exp, true)
end

-- repl
local repl_env = {['+'] = function(a,b) return a+b end,
                  ['-'] = function(a,b) return a-b end,
                  ['*'] = function(a,b) return a*b end,
                  ['/'] = function(a,b) return math.floor(a/b) end}
function rep(str)
    return PRINT(EVAL(READ(str),repl_env))
end

if #arg > 0 and arg[1] == "--raw" then
    readline.raw = true
end

while true do
    line = readline.readline("user> ")
    if not line then break end
    xpcall(function()
        print(rep(line))
    end, function(exc)
        if exc then
            if types._malexception_Q(exc) then
                exc = printer._pr_str(exc.val, true)
            end
            print("Error: " .. exc)
            print(debug.traceback())
        end
    end)
end

File: impls/objc/step2_eval.m
#import <Foundation/Foundation.h>

#import "mal_readline.h"
#import "types.h"
#import "reader.h"
#import "printer.h"

// read
NSObject *READ(NSString *str) {
    return read_str(str);
}

// eval
NSObject *EVAL(NSObject *ast, NSDictionary *env) {
    // NSLog(@"EVAL: %@ (%@)", _pr_str(ast, true), env);
    if ([ast isMemberOfClass:[MalSymbol class]]) {
        if ([env objectForKey:ast]) {
            return env[ast];
        } else {
            @throw [NSString stringWithFormat:@"'%@' not found", ast];
        }
    } else if ([ast isKindOfClass:[MalVector class]]) {
        NSMutableArray *newLst = [NSMutableArray array];
        for (NSObject * x in (NSArray *)ast) {
            [newLst addObject:EVAL(x, env)];
        }
        return [MalVector fromArray:newLst];
    } else if ([ast isKindOfClass:[NSDictionary class]]) {
        NSMutableDictionary *newDict = [NSMutableDictionary dictionary];
        for (NSString * k in (NSDictionary *)ast) {
            newDict[k] = EVAL(((NSDictionary *)ast)[k], env);
        }
        return newDict;
    } else if (! [ast isKindOfClass:[NSArray class]]) {
        return ast;
    }

    // apply list
    NSArray * alst = (NSArray *)ast;
    if ([alst count] == 0) {
        return ast;
    }
    id el0 = EVAL(alst[0], env);
    NSObject * (^ f)(NSArray *) = el0;
    NSMutableArray * args = [NSMutableArray array];
    for (int i = 1; i < [alst count]; i++) {
        [args addObject:EVAL(alst[i], env)];
    }
    return f(args);
}

// print
NSString *PRINT(NSObject *exp) {
    return _pr_str(exp, true);
}

// REPL
NSString *REP(NSString *line, NSDictionary *env) {
    return PRINT(EVAL(READ(line), env));
}

int main () {
    NSDictionary * repl_env = @{
        @"+": ^(NSArray *args){
            return [NSNumber numberWithInt:[args[0] intValue] + [args[1] intValue]];
        },
        @"-": ^(NSArray *args){
            return [NSNumber numberWithInt:[args[0] intValue] - [args[1] intValue]];
        },
        @"*": ^(NSArray *args){
            return [NSNumber numberWithInt:[args[0] intValue] * [args[1] intValue]];
        },
        @"/": ^(NSArray *args){
            return [NSNumber numberWithInt:[args[0] intValue] / [args[1] intValue]];
        },
        };

    // Create an autorelease pool to manage the memory into the program
    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
    // If using automatic reference counting (ARC), use @autoreleasepool instead:
//    @autoreleasepool {

    while (true) {
        char *rawline = _readline("user> ");
        if (!rawline) { break; }
        NSString *line = [NSString stringWithUTF8String:rawline];
        if ([line length] == 0) { continue; }
        @try {
            printf("%s\n", [[REP(line, repl_env) description] UTF8String]);
        } @catch(NSString *e) {
            printf("Error: %s\n", [e UTF8String]);
        } @catch(NSException *e) {
            if ([[e name] isEqualTo:@"ReaderContinue"]) { continue; }
            printf("Exception: %s\n", [[e reason] UTF8String]);
        }
    }

    [pool drain];

//    }
}
